util/rom-ext-manifest-generator.py

```diff
@@ -0,0 +1,106 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import copy
+from pathlib import Path
+
+import hjson
+from mako.template import Template
+
+from topgen.c import MemoryRegion, Name
+
+DESC = """TODO rom-ext-manifest-gen"""
+
+USAGE = '''
+  rom-ext-manifest-generator [options]
+  rom-ext-manifest-generator [options] <input>
+  rom-ext-manifest-generator (-h | --help)
+'''
+
+# This file is $REPO_TOP/util/rom-ext-manifest-generator.py, so it takes two
+# parent() calls to get back to the top.
+REPO_TOP = Path(__file__).resolve().parent.parent
+
+
+# Adds `size_in_bytes` entry to every parsed ROM_EXT manifest field.
+# fields[N]    =>    fields[N]
+#   - ...              - ...
+#   - size             - size
+#                      - size_in_bytes
+def add_size_in_bytes(fields):
+    for field in fields:
+        field['size_in_bytes'] = str(int(field['size'], 0) // 8)
+
+
+# Adds `offset_in_bytes` entry to every parsed ROM_EXT manifest field.
+# fields[N]    =>    fields[N]
+#   - ...              - ...
+#   - size_in_bytes    - size_in_bytes
+#                      - offset_in_bytes
+def add_offsets_in_bytes(fields):
+    accumulated_offset = 0x0
+
+    for field in fields:
+        field['offset'] = str(accumulated_offset)
+        accumulated_offset += int(field['size_in_bytes'], 0)
+
+
+# Removes `reserved` entries to from parsed ROM_EXT manifest fields.
+# field0, reserved, field1, ... => field0, field1, ...
+# This is simple way to prevent `generate_cheader` from generating
+# defines for metadata.
+def strip_reserved(fields):
+    for field in fields:
+        if field['name'] == "reserved":
+            fields.remove(field)
+
+
+# Generates C header file from the `template_file`. It produces a dictionary
+# where `fields`, where a field name is a key, and the `MemoryRegion` object
+# is a value. This dictionary is used in the `template_file`, and the resulting
+# header file is placed in the `output_dir`.
+def generate_cheader(fields, template_file, output_dir):
+    base_name = Name.from_snake_case("ROM_EXT")
+
+    items = [(field['name'],
+              MemoryRegion(base_name + Name.from_snake_case(field['name']),
+                           field['offset'], field['size_in_bytes']))
+             for field in fields]
+
+    with template_file.open('r') as f:
+        template = Template(f.read())
+
+    header = template.render(items=items)
+
+    output = output_dir / 'manifest.h'
+    with output.open('w') as f:
+        f.write(header)
+
+    print('Template sucessfuly written to {}.'.format(str(output)))
+
+
+def main():
+    # Default defines
+    rom_ext_dir = REPO_TOP / 'sw/device/rom_exts'
+    manifest_hjson_file = rom_ext_dir / 'manifest.hjson'
+    manifest_cheader_template_file = rom_ext_dir / 'manifest.h.tpl'
+    # TODO - manifest_asm_template_file
+
+    with manifest_hjson_file.open('r') as hjson_file:
+        obj = hjson.loads(hjson_file.read())
+
+    fields = copy.deepcopy(obj['fields'])
```

Comment:
`obj['fields']` is never used again; is there a reason for the deep copy?

Reply:
The idea was that we would have an original list of fields that could be extrapolated differently for different types of generation if required - hence the deepcopy. At the moment there is only C header generation, so it is not really needed.

Reply:
I think it would be *much* nicer to avoid explicit copying. It's quite an anti-pattern in languages higher-level than C. Here, I think you could simplify the code as well as getting rid of all the deepcopy nonsense with something like this:
```python
def add_size_and_offsets(fields):
    """Adds `size_in_bytes` and `offset` entries to fields.

    Each field is expected to have a "size" entry, giving the size in bits.
    From this, we derive the size in bytes and accumulate an offset in bytes
    starting at zero.

    """

    ret = []
    offset = 0
    for field in fields:
        assert field['size'] % 8 == 0

        new_field = field.copy()
        size_in_bytes = field['size'] // 8

        new_field['size_in_bytes'] = size_in_bytes
        new_field['offset'] = hex(offset)
        ret.append(new_field)

        offset += size_in_bytes

    return ret

...

    c_fields = add_size_and_offsets(obj['fields'])
    generate_cheader(c_fields, input_dir, output_dir)

```
Incidentally, the docstring comment for `add_offsets_in_bytes` has a different field name from the code. I matched the code.

Also, I'm not a fan of storing a string as an offset: is there a reason to render it here rather than in the template?

Reply:
(Also: I think the code would be better without the copying and helper functions at all: see [here](https://github.com/lowRISC/opentitan/pull/3576#discussion_r500328377)).

Reply:
Done.

Reply:
Followed the suggestion from your other comment.